[RFC] Introduce initial PAN support for OP-TEE OS

seonghp opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->
Privileged Access Never (PAN) is a part of ARMv8.1 extension that restricts accesses to unprivileged memory from privileged mode in order to prevent unintended accesses to potentially malicious
memory.

Previously, OP-TEE implementation was not ready to apply strict PAN policy due to missing user-access function uses, etc.

After spotting out all outstanding missing unprivileged access functions (adding few more such functions if needed), fine-grained PAN support became possible.

Note that while it is typical to copy contents in the user-space into the kernel space before accessing it, due to potential performance impact, this patch does not allocate such bounce buffers for cryptographic operations that accesses large user-space input and output data buffers

jenswi-linaro commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -143,6 +148,13 @@
+		mrs	x0, sctlr_el1
+		bic	x0, x0, #SCTLR_SPAN
+		msr	sctlr_el1, x0
+		.inst	0xd500409f | (0x1 << 8) /* Set PSTATE.PAN */
```
Please let the assembler translate the instruction into machine code instead.

seonghp replied:
Unfortunately the assembler is not able to translate `msr pan, x0`, etc. for now. The compiler would output error like `core/arch/arm/kernel/entry_a64.S:194: Error: selected processor does not support PSTATE field name 'pan'`

jenswi-linaro replied:
Please add
```
        .macro write_pan reg
        /* msr pan, \reg */
        msr     S3_0_c4_c2_3, \reg
        .endm

        .macro write_pan_enable
        /* msr pan, #1 */
        msr     S0_0_c4_c1_4, xzr
        .endm

        .macro write_pan_disable
        /* msr pan, #0 */
        msr     S0_0_c4_c0_4, xzr
        .endm
```
to `core/arch/arm/include/arm64_macros.S` and use those macros instead.

jenswi-linaro commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -173,6 +185,14 @@ FUNC _start , :
 	isb
 
+#ifdef CFG_PAN
+	read_feat_pan x0
```
I suppose this could go into the `init_pan` macro too.

seonghp replied:
moved into `init_pan`

jenswi-linaro reviewed (commented):
Comments for the commit "arm64: Introduce permissive PAN implementation"

jenswi-linaro commented on core/kernel/user_access.c:
```diff
@@ -37,7 +38,7 @@ TEE_Result copy_from_user(void *kaddr, const void *uaddr, size_t len)
 TEE_Result copy_to_user(void *uaddr, const void *kaddr, size_t len)
 {
-	uint32_t flags = TEE_MEMORY_ACCESS_WRITE | TEE_MEMORY_ACCESS_ANY_OWNER;
+	uint32_t flags = TEE_MEMORY_ACCESS_READ | TEE_MEMORY_ACCESS_ANY_OWNER;
```
What if the target user address isn't writeable? Then well crash OP-TEE core, right?

seonghp replied:
I reverted this change. This was necessary previously, but now no longer needed

jenswi-linaro commented on core/include/kernel/user_access.h:
```diff
@@ -33,6 +33,21 @@ static inline TEE_Result copy_from_user(void *kaddr __unused,
 TEE_Result copy_to_user(void *uaddr, const void *kaddr, size_t len);
 
+#define get_user(_x, _p) ({						\
+	TEE_Result __res;						\
```
The coding style requires local variables to be initialized when declared, `0` is fine.

etienne-lms replied:
Use upper case names for macro: `GET_USER()`, `PUT_USER()`.
Implementation should check type of `_x` and `_p` match is size.
Caller shall check the return values.

seonghp replied:
Added initial value

seonghp replied:
Added compile-time assertion

jenswi-linaro commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -51,7 +51,7 @@ TEE_Result ldelf_syscall_map_zi(vaddr_t *va, size_t num_bytes, size_t pad_begin,
 		return TEE_ERROR_OUT_OF_MEMORY;
 	res = vm_map_pad(uctx, va, num_bytes, prot, vm_flags,
-			 mobj, 0, pad_begin, pad_end, 0);
+			 mobj, 0, pad_begin, pad_end, 0, true);
```
`ldelf` is the only user space component that calls `vm_map_pad()` and friends with a user space pointer `vaddr_t *va` in the argument. I'd rather read and write the `va` pointer in `ldelf_syscall_map_zi()` and `ldelf_syscall_map_bin()` as needed.

etienne-lms replied:
not addressed.
Discard changes in `vm_map_pad()` and `vm_map()` and instead use something like:
```c
	vaddr_t _va = 0;

	(...)

	res = GET_USER(_va, va);
	if (!res)
		res = vm_map_pad(uctx, _va, num_bytes, prot, vm_flags,
				 mobj, 0, pad_begin, pad_end, 0);
```
Ditto in `ldelf_syscall_map_bin()`.

jenswi-linaro commented on core/arch/arm/kernel/arch_scall.c:
```diff
@@ -94,9 +95,14 @@ static void save_panic_regs_a32_ta(struct thread_specific_data *tsd,
 static void save_panic_regs_a64_ta(struct thread_specific_data *tsd,
 				   uint64_t *pushed)
 {
+	uint64_t x29, elr;
```
Please declare one local variable per line and initialize them.

seonghp replied:
updated

jenswi-linaro commented on core/include/kernel/ldelf_syscalls.h:
```diff
@@ -31,6 +31,11 @@ TEE_Result ldelf_syscall_map_bin(vaddr_t *va, size_t num_bytes,
 				 unsigned long flags);
 TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
```
We can remove `ldelf_syscall_copy_from_bin()` and its syscall entry if it's not used any longer. `ldelf` is always updated in sync with OP-TEE core so there's no need for a stable ABI.

seonghp replied:
`ldelf_syscall_copy_from_bin()` is still used by `copy_remapped_to`

jenswi-linaro replied:
OK

jenswi-linaro reviewed (commented):
Comment for "misc: allow to_user_mode_ctx to return NULL"

No, this is done like this on purpose. The pointer should be checked first with `is_user_mode_ctx()` unless it's known to be good.

seonghp commented:
Thank you @jenswi-linaro for your invaluable reviews. I will handle them soon

etienne-lms commented on core/kernel/user_access.c:
```diff
@@ -55,21 +80,78 @@ TEE_Result copy_from_user_private(void *kaddr, const void *uaddr, size_t len)
+	return res;
+}
+
+size_t clear_user(void *uaddr, size_t n)
```
why not returning a `TEE_Result`?

seonghp replied:
Function type of `clear_user` was taken from other operating systems that have similar functions. updated the function to return `TEE_Result`

etienne-lms commented on core/kernel/user_access.c:
```diff
@@ -55,21 +80,78 @@ TEE_Result copy_from_user_private(void *kaddr, const void *uaddr, size_t len)
+	if (!*p)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if ((res = copy_from_user(*p, src, len)) != TEE_SUCCESS)	 {
```
2 steps:
```
	res = copy_from_user(*p, src, len);
	if (res) {
		...

seonghp replied:
updated

etienne-lms commented on core/tee/tee_svc_storage.c:
```diff
@@ -642,6 +643,7 @@ TEE_Result syscall_storage_next_enum(unsigned long obj_enum,
 	TEE_Result res = TEE_SUCCESS;
 	struct tee_obj *o = NULL;
 	uint64_t l = 0;
+	struct utee_object_info buf = { 0, };
```
`= { }`

seonghp replied:
updated

etienne-lms commented on mk/config.mk:
```diff
@@ -943,6 +943,12 @@ ifeq (y-y,$(CFG_WITH_PAGER)-$(CFG_MEMTAG))
 
+# Privileged Access Never (PAN, part of the ARMv8.1 Extensiosn) can be
+# used to restrict accesses to unprivileged memory from privileged mode.
+CFG_PAN ?= n
```
PAN is arm specific. Rename to `CFG_ARM_PAN`?

seonghp replied:
I chose config name after `CFG_MEMTAG`, which is probably only available on ARM. That being said, if you think `CFG_ARM_PAN` if more appropriate, I will rename the config

etienne-lms replied:
@jenswi-linaro, @jforissier: `CFG_PAN` or `CFG_ARM_PAN`? or maybe generic `CFG_PRIVILEGED_ACCESS_NEVER` (that Arm archs can implement using PAN CPU feature).

jforissier replied:
`CFG_PAN` sounds good to me, PAN is PAN and short names are better. In addition, the macro is used only in Arm-specific files so there is no need to make it sound more general than what it means.

etienne-lms replied:
PAN is PAN? IDTS: https://en.wikipedia.org/wiki/Pan
That said, I have no strong opinion on this by that i don't really like abusive use of acronyms that can be confusing IMHO, FWIW.

etienne-lms replied:
I suggest to rephrase "PAN, part of the ARMv8.1 Extensiosn" to "PAN in part of the ARMv8.1 Extensiosn" since `CFG_PAN` is not Arm architecture specific.

etienne-lms commented on ldelf/ta_elf.c:
```diff
@@ -874,17 +874,15 @@ static void populate_segments(struct ta_elf *elf)
-							   elf->handle, offset);
-				if (res)
-					err(res, "sys_copy_from_ta_bin");
+					err(res, "sys_map_zi_and_copy_from_bin");
```
s/sys_map_zi_and_copy_from_bin/sys_map_zi_and_copy_from_bin/

seonghp replied:
maybe you meant `s/sys_map_zi_and_copy_from_bin/sys_map_zi_and_copy_from_ta_bin`? updated

etienne-lms replied:
yes, sorry. thanks.

etienne-lms commented on ldelf/sys.c:
```diff
@@ -72,6 +72,16 @@ TEE_Result sys_copy_from_ta_bin(void *dst, size_t num_bytes, uint32_t handle,
 	return _ldelf_cp_from_bin(dst, offs, num_bytes, handle);
 }
 
+TEE_Result sys_map_zi_and_copy_from_ta_bin(vaddr_t *va, size_t memsz,
```
A bit a inline description would be welcome. Maybe in the header file.

seonghp replied:
left inline comments

etienne-lms commented on core/tee/tee_svc.c:
```diff
@@ -815,7 +820,7 @@ static TEE_Result tee_svc_update_out_param(
 	size_t n;
 	uint64_t *vals = usr_param->vals;
-	size_t sz = 0;
+	size_t sz = 0, in_sz;
```
1 variable defintinio per line.
Add an initialization value.

seonghp replied:
updated

seonghp commented on core/arch/arm/kernel/entry_a64.S:
```diff
@@ -173,6 +185,14 @@ FUNC _start , :
+	cmp	x0, #0
+	b.eq	3333f
+	init_pan
+3333:
```
updated

seonghp commented on core/include/kernel/user_access.h:
```diff
@@ -42,4 +42,18 @@ static inline void *uref_to_kaddr(uint32_t uref)
 	return (void *)uref_to_vaddr(uref);
 }
 
+#define ENTER_USER_ACCESS	\
```
moved PAN-specific code into arm64.h

seonghp commented on core/include/kernel/user_access.h:
```diff
@@ -42,4 +42,18 @@ static inline void *uref_to_kaddr(uint32_t uref)
+	".inst	0xd500409f | (1 << 8);"
+
+#ifdef CFG_PAN
+#define enter_user_access()	asm volatile(ENTER_USER_ACCESS)
```
I created `user_access_arch.h` and located implementation of `enter_user_access()` and `exit_user_access()` there. I will check again for the barrier part

seonghp replied:
According to the Arm Architecture Reference Manual for A-profile architecture C5.1.3,
> Writes to PSTATE occur in program order without the need for additional synchronization.

IIUC barrier is not required

seonghp commented:
> Comment for "misc: allow to_user_mode_ctx to return NULL"
> 
> No, this is done like this on purpose. The pointer should be checked first with `is_user_mode_ctx()` unless it's known to be good.

I reverted this change and updated `check_access`'s logic to test `s->ctx` with `is_user_mode_ctx()`.

etienne-lms commented on core/arch/arm/include/kernel/user_access_arch.h:
```diff
@@ -18,4 +18,26 @@ static inline void enter_user_access(void) {}
+#define put_user(_x, _p) ({						\
+	TEE_Result __res = TEE_SUCCESS;					\
+	__typeof(*(_p)) __x = (_x);					\
+	COMPILE_TIME_ASSERT(sizeof(_x) == sizeof(*_p));			\
```
s/*_p/*(_p)/
use `static_assert()` instead of `COMPILE_TIME_ASSERT()`.
add an empty line after local variable definition.

By the way, @jforissier, @jenswi-linaro, would you be ok we create a P-R to remove `COMPILE_TIME_ASSERT()`?

seonghp replied:
switched to `static_assert()`

etienne-lms commented on core/arch/arm/include/kernel/user_access_arch.h:
```diff
@@ -18,4 +18,26 @@ static inline void enter_user_access(void) {}
+	__res;								\
+})
+
+#define put_user(_x, _p) ({						\
```
prefer upper case for macro names.

seonghp replied:
capitalized the macro

etienne-lms reviewed (commented):
Dear reviewers, are you ok to ask fixup commits to be squash to get an clear patch series?

etienne-lms commented on core/kernel/user_access.c:
```diff
@@ -29,8 +49,9 @@ TEE_Result copy_from_user(void *kaddr, const void *uaddr, size_t len)
-	if (!res)
+	if (!res) {
 		memcpy(kaddr, uaddr, len);
+	}
```
commit "core: add more user access functions" should not add these braces. Should be done from commit "core: Apply finer-grained PAN".

etienne-lms commented on core/kernel/user_access.c:
```diff
@@ -68,8 +91,57 @@ TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len)
+
+TEE_Result clear_user(void *uaddr, size_t n)
+{
+	uint32_t flags = TEE_MEMORY_ACCESS_READ | TEE_MEMORY_ACCESS_ANY_OWNER;
```
replace `TEE_MEMORY_ACCESS_READ` with `TEE_MEMORY_ACCESS_WRITE`

etienne-lms commented on core/kernel/user_access.c:
```diff
@@ -68,8 +91,57 @@ TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len)
+
+	memset(uaddr, 0, n);
+
+	return TEE_SUCCESS;
```
Shorten with:
```c
	res = check_access(flags, uaddr, n);
	if (!res)
		memset(uaddr, 0, n);
	return res;

etienne-lms commented on core/include/kernel/user_access.h:
```diff
@@ -34,6 +34,20 @@ static inline TEE_Result copy_from_user(void *kaddr __unused,
 TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len);
 TEE_Result copy_to_user(void *uaddr, const void *kaddr, size_t len);
 
+TEE_Result clear_user(void *uaddr, size_t n);
```
Add an inline fucntion so one get the diff versus a simple `memset(uaddr, 0, n)` instruction.

etienne-lms replied:
Discard my comment. Looking at the other function below and above this line, I guess it not needed to add a description specifically for this function. Sorry for the noise.

etienne-lms commented on core/include/kernel/user_access.h:
```diff
@@ -34,6 +34,20 @@ static inline TEE_Result copy_from_user(void *kaddr __unused,
+ * @len:    Length of the user buffer to be duplicated.
+ * @p:      Holds duplicated kernel buffer on success, or NULL on failure
+ * @returns TEE_SUCCESS on success, and TEE_ERROR_OUT_OF_MEMORY or
+ *          TEE_ERROR_ACCESS_DENIEND on error.
```
s/TEE_ERROR_ACCESS_DENIEND/TEE_ERROR_ACCESS_DENIED/

State that the memory is allocated with `malloc()` or like and hence can be freed with `free()`.

seonghp replied:
fixed typo and added comments

etienne-lms replied:
the wrong commit has the fix: please squash the fix in this very commit.

etienne-lms commented on core/kernel/user_access.c:
```diff
@@ -68,8 +91,57 @@ TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len)
+	if (res) {
+		free(*p);
+		*p = NULL;
+		return res;
```
can remove this instruction, already done 3 lines below

etienne-lms commented on core/arch/arm/kernel/arch_scall.c:
```diff
@@ -94,9 +95,15 @@ static void save_panic_regs_a32_ta(struct thread_specific_data *tsd,
+	uint64_t elr = 0;
+
+	GET_USER(x29, &pushed[0]);
+	GET_USER(elr, &pushed[1]);
```
check return value from `GET_USER()`. If I understand correctly the implementation, you can set `tsd->abort_regs.x29` and `tsd->abort_regs.elr` to 0 upon failure.

@jenswi-linaro, @jforissier: a expert advise is welcome here. Is user land x1 register always trustfull in this case?

seonghp replied:
added some return value checks

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -446,6 +447,90 @@ TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
+	if (!mobj)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	GET_USER(vaddr, va);
```
check `GET_USER()` return value?

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -446,6 +447,90 @@ TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
+	 */
+	vm_set_ctx(uctx->ts_ctx);
+
+	put_user(vaddr, va);
```
s/put_user/PUT_USER/

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -446,6 +447,90 @@ TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
+
+	put_user(vaddr, va);
+
+	return res;
```
explicit `return TEE_SUCCESS` would be better.

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -446,6 +447,90 @@ TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
+	/*
+	 * The context currently is active set it again to update
+	 * the mapping.
+	 */
```
fits in a single line:
```c
	/* The context currently is active set it again to update the mapping */
```

seonghp replied:
updated

etienne-lms commented on core/kernel/ts_manager.c:
```diff
@@ -62,7 +62,13 @@ struct ts_session *ts_get_calling_session(void)
 struct ts_session *ts_get_current_session_may_fail(void)
 {
-	return TAILQ_FIRST(&thread_get_tsd()->sess_stack);
+	struct thread_specific_data* tsd;
```
= NULL

etienne-lms commented on core/kernel/thread.c:
```diff
@@ -499,6 +499,14 @@ void thread_init_core_local_pauth_keys(void)
 
+struct thread_specific_data * __noprof thread_get_tsd_may_fail(void)
+{
+	short int thread_id = thread_get_id_may_fail();
```
add an empty line below

etienne-lms commented on core/kernel/ldelf_loader.c:
```diff
@@ -356,26 +361,37 @@ TEE_Result ldelf_dlopen(struct user_mode_ctx *uctx, TEE_UUID *uuid,
 	uint32_t panic_code = 0;
 	uint32_t panicked = 0;
 	struct ts_session *sess = NULL;
+	struct dl_entry_arg dl_entry_arg = { 0, };
```
`= { }`

etienne-lms commented on core/kernel/ldelf_loader.c:
```diff
@@ -406,30 +424,50 @@ TEE_Result ldelf_dlsym(struct user_mode_ctx *uctx, TEE_UUID *uuid,
-	size_t len = strnlen(sym, maxlen);
+	size_t len = strnlen_user(sym, maxlen);
 	struct ts_session *sess = NULL;
+	struct dl_entry_arg *dl_entry_arg;
```
= NULL

etienne-lms commented on core/kernel/ldelf_loader.c:
```diff
@@ -452,5 +490,12 @@ TEE_Result ldelf_dlsym(struct user_mode_ctx *uctx, TEE_UUID *uuid,
+		free(dl_entry_arg);
+out:
+	if (uuid_kbuf)
+		free(uuid_kbuf);
```
no need to test `dl_entry_arg` and `uuid_kbuf`. It is allowed to call `free(NULL)`.

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -133,17 +134,21 @@ TEE_Result ldelf_syscall_open_bin(const TEE_UUID *uuid, size_t uuid_size,
 		sess->user_ctx = sys_ctx;
 	}
 
+	res = memdup_user(uuid, uuid_size, (void **)&uuid_kbuf);
```
Remove `vm_check_access_rights()` at line 113 as `memdup_user()` already takes care of this sanity test.

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -181,13 +186,18 @@ TEE_Result ldelf_syscall_open_bin(const TEE_UUID *uuid, size_t uuid_size,
 		goto err_oom;
-	*handle = h;
+
+	PUT_USER(h, handle);
```
Remove `vm_check_access_rights()` at line 120 as `PUT_USER()` already takes care of this sanity test.

etienne-lms commented on core/mm/vm.c:
```diff
@@ -612,7 +616,7 @@ TEE_Result vm_remap(struct user_mode_ctx *uctx, vaddr_t *new_va, vaddr_t old_va,
 			res = umap_add_region(&uctx->vm_info, r, 0, 0, 0);
 		} else {
-			r->va = *new_va;
+			GET_USER(r->va, new_va);
```
check return values

etienne-lms commented on core/mm/vm.c:
```diff
@@ -664,7 +668,7 @@ TEE_Result vm_remap(struct user_mode_ctx *uctx, vaddr_t *new_va, vaddr_t old_va,
 
 	vm_set_ctx(uctx->ts_ctx);
-	*new_va = r_first->va;
+	PUT_USER(r_first->va, new_va);
```
check return value

seonghp commented on core/arch/arm/include/kernel/user_access_arch.h:
```diff
@@ -18,4 +18,26 @@ static inline void enter_user_access(void) {}
 static inline void exit_user_access(void) {}
 #endif /* CFG_PAN */
 
+#ifdef ARM64
```
the problem is that if the upper implementation is used for ARM32, then the compile-time size equality check would fail in many places.

seonghp commented:
Thank you @etienne-lms for kind and detailed reviews. I will handle them soon

seonghp reviewed (commented):
addressed part of the comments. This is currently wip

seonghp reviewed (commented):
Thank you @etienne-lms for the reviews!

I took another iteration. IMHO most of the comments are handled, except for those regarding `GET_USER()` and `PUT_USER()`, and naming of the `enter_user_access()` and `exit_user_access()`.

Implementing multi-platform `GET_USER()` and `PUT_USER()` was kinda tricky due to type size mismatch, yet I will prepare fixes soon.

Regarding the naming of `enter_user_access`, I'd prefer name that are architecture-agnostic, rather than that are specific to ARM architecture (e.g., `enable_pan`). Because RISC-V also has hardware feature called SSTATUS.SUM (which stands for 'permit Supervisor User Memory access') that is almost equivalent to PAN, and I believe architecture-agnostic naming shall be preferred so that RISC-V can reuse the interface by implementing that function in the future. Some names that I could came up with were `(dis)allow_user_access()` or `(dis)allow_unprivileged_access()`. Do you have more suggestions?

etienne-lms commented on core/kernel/ldelf_loader.c:
```diff
@@ -406,30 +430,50 @@ TEE_Result ldelf_dlsym(struct user_mode_ctx *uctx, TEE_UUID *uuid,
+		goto out;
+	}
+
+	dl_entry_arg = (struct dl_entry_arg*)malloc(len_arg_with_sym);
```
cast not needed

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -50,11 +52,26 @@ TEE_Result ldelf_syscall_map_zi(vaddr_t *va, size_t num_bytes, size_t pad_begin,
 	mobj_put(mobj);
 
-	return res;
+	res = PUT_USER(_va, va);
```
test `res` value got from line 60 before line 64.

etienne-lms commented on core/kernel/ldelf_loader.c:
```diff
@@ -452,5 +496,10 @@ TEE_Result ldelf_dlsym(struct user_mode_ctx *uctx, TEE_UUID *uuid,
 			*val = arg->dlsym.val;
 	}
 
+out_free_dl_entry_arg:
```
note here you can simplify by dropping branch label `out_free_dl_entry_arg` and only use `out` as branch label in error cases above. This is permitted as `free(NULL)` is valid:
```c
out:
	free(dl_entry_arg);
	free(uuid_kbuf);

	return res;
```

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -50,11 +52,26 @@ TEE_Result ldelf_syscall_map_zi(vaddr_t *va, size_t num_bytes, size_t pad_begin,
+		if (vm_unmap(uctx, _va, num_bytes))
+			panic();
+
+		vm_set_ctx(uctx->ts_ctx);
```
is it needed to call `vm_set_ctx(uctx->ts_ctx)`? i don't think so. Unmap instruction from line 66 is already effective once the function return with success.

etienne-lms replied:
or maybe i'm wrong. @jenswi-linaro?

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -180,13 +187,20 @@ TEE_Result ldelf_syscall_open_bin(const TEE_UUID *uuid, size_t uuid_size,
+	if (res)
+		goto err;
+
+	free(uuid_kbuf);
```
could free `uuid_kbuf` straight from line 173:
```patch
 	binh = calloc(1, sizeof(*binh));
 	if (!binh)
 		goto err_oom;
 
 	if (is_user_ta_ctx(sess->ctx) || is_stmm_ctx(sess->ctx)) {
 		...
 	} else if (is_sp_ctx(sess->ctx)) {
 		...
 	} else {
 		res = TEE_ERROR_ITEM_NOT_FOUND;
 	}
+	free(uuid_kbuf);

etienne-lms commented on core/mm/vm.c:
```diff
@@ -605,7 +610,10 @@ TEE_Result vm_remap(struct user_mode_ctx *uctx, vaddr_t *new_va, vaddr_t old_va,
 			res = umap_add_region(&uctx->vm_info, r, 0, 0, 0);
 		} else {
-			r->va = *new_va;
+			res = GET_USER(r->va, new_va);
```
could be done once for all at function entry.

etienne-lms commented on core/mm/vm.c:
```diff
@@ -657,7 +665,10 @@ TEE_Result vm_remap(struct user_mode_ctx *uctx, vaddr_t *new_va, vaddr_t old_va,
+
+	res = PUT_USER(r_first->va, new_va);
+	if (res)
+		goto err_restore_map;
```
I don't think `err_restore_map` will work at this stage as `fobj` is lost.
I would rather swap the above instructions:
```
	res = PUT_USER(r_first->va, new_va);
	if (res)
		goto err_restore_map;

	fobj_put(fobj);

	vm_set_ctx(uctx->ts_ctx);

seonghp replied:
thank you for the catch. updated

etienne-lms commented on core/pta/system.c:
```diff
@@ -50,9 +53,16 @@ static TEE_Result system_rng_reseed(uint32_t param_types,
+
+	free(seed);
+
+	return res;
```
`return TEE_SUCCESS`

seonghp replied:
updated

etienne-lms commented on core/pta/system.c:
```diff
@@ -103,14 +114,26 @@ static TEE_Result system_derive_ta_unique_key(struct user_mode_ctx *uctx,
+out_free_subkey:
+	free_wipe(subkey);
+out:
+	free_wipe(data);
```
simplify:
```c
out:
	free_wipe(subkey);
	free_wipe(data);

seonghp replied:
update

etienne-lms commented on core/tee/tee_pobj.c:
```diff
@@ -64,14 +65,36 @@ TEE_Result tee_pobj_get(TEE_UUID *uuid, void *obj_id, uint32_t obj_id_len,
+	res = memdup_user(obj_id, obj_id_len, &obj_id_kbuf);
+	switch (res) {
+	case TEE_SUCCESS:
+		break;
```
explicit `obj_id_from_user = true;` here would be better IMHO

seonghp replied:
updated

etienne-lms commented on core/tee/tee_pobj.c:
```diff
@@ -127,6 +150,8 @@ TEE_Result tee_pobj_get(TEE_UUID *uuid, void *obj_id, uint32_t obj_id_len,
 	if (res != TEE_SUCCESS)
 		*obj = NULL;
+	if (obj_id_from_user && obj_id_kbuf != NULL)
+		free(obj_id_kbuf);
```
```c
if (obj_id_from_user)
	free(obj_id_kbuf)

etienne-lms commented on core/tee/tee_supp_plugin_rpc.c:
```diff
@@ -74,5 +80,8 @@ TEE_Result tee_invoke_supp_plugin_rpc(const TEE_UUID *uuid, uint32_t cmd,
 		thread_rpc_free_payload(mobj);
 
+	if (uuid_kbuf)
+		free(uuid_kbuf);
```
```patch
-	if (uuid_kbuf)
-		free(uuid_kbuf);
+	free(uuid_kbuf);

seonghp replied:
updated

etienne-lms commented on core/tee/tee_svc.c:
```diff
@@ -828,7 +838,10 @@ static TEE_Result tee_svc_update_out_param(
 			sz = param->u[n].mem.size;
-			if (tmp_buf_va[n] && sz <= vals[n * 2 + 1]) {
+
+			GET_USER(in_sz, &vals[n * 2 + 1]);
```
test return value?

etienne-lms commented on core/tee/tee_svc.c:
```diff
@@ -845,13 +858,16 @@ static TEE_Result tee_svc_update_out_param(
 				}
 			}
-			usr_param->vals[n * 2 + 1] = sz;
+			PUT_USER(sz, &usr_param->vals[n * 2 + 1]);
```
test return value?

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -695,14 +696,17 @@ static bool op_u32_from_binary_helper(uint32_t *v, const uint8_t *data,
+		return res;
 	key->key_size = size;
-	return TEE_SUCCESS;
+	return res;
```
`return TEE_SUCCESS` is fine here

etienne-lms replied:
or: 
```c
	res = copy_from_user(key + 1, buffer, size);
	if (!res)
		key->key_size = size;

	return res;

etienne-lms reviewed (commented):
I wonder if return value of `GET_USER()` and `PUT_USER()` should be tested everytime the function is used. It's a bit confusing.
Sorry for flooding with comments.

etienne-lms commented on core/kernel/ldelf_loader.c:
```diff
@@ -452,5 +496,10 @@ TEE_Result ldelf_dlsym(struct user_mode_ctx *uctx, TEE_UUID *uuid,
 			*val = arg->dlsym.val;
```
`PUT_USER()` for `*val = arg->dlsym.val;` ?
`

etienne-lms replied:
so I would prefer you add an inline comment stating the `PUT_USER()` is not needed here for that reasonb, otherwise it looks as if we missed something.

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -373,16 +396,22 @@ TEE_Result ldelf_syscall_map_bin(vaddr_t *va, size_t num_bytes,
+
+		res = GET_USER(_va, va);
+		if (res)
+			goto err;
```
`mobj_put()`

etienne-lms commented on core/kernel/user_ta.c:
```diff
@@ -54,7 +54,13 @@ static void init_utee_param(struct utee_params *up,
+	 * is uint32_t. Hence, to avoid unexpected behaviors
+	 */
+	clear_user(&up->types, sizeof(up->types));
+	copy_to_user(&up->types, &p->types, sizeof(p->types));
```
test resturn value?

seonghp replied:
shall I change the function return type?

etienne-lms replied:
yes, i think it needs to

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3487,6 +3578,15 @@ TEE_Result syscall_cryp_derive_key(unsigned long state,
 	const struct tee_cryp_obj_type_props *type_props = NULL;
 	TEE_Attribute *params = NULL;
 	size_t alloc_size = 0;
+	size_t i;
```
= 0

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3487,6 +3578,15 @@ TEE_Result syscall_cryp_derive_key(unsigned long state,
 	TEE_Attribute *params = NULL;
 	size_t alloc_size = 0;
+	size_t i;
+	void **bufs = NULL; /* bufs[N] is used as a bounce buffer for
```
move inline comment to a specif line

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3487,6 +3578,15 @@ TEE_Result syscall_cryp_derive_key(unsigned long state,
+			       params[N].ref, if needed */
+
+	bufs = malloc(sizeof(void *) * param_count);
+	if (!bufs)
```
`bufs` must be free in error cases below

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3594,17 +3697,23 @@ TEE_Result syscall_cryp_derive_key(unsigned long state,
 			goto out;
 		}
 
+		res = memdup_user(params[0].content.ref.buffer, params[0].content.ref.length, &bufs[0]);
```
add a line break

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3785,27 +3899,30 @@ TEE_Result syscall_cryp_derive_key(unsigned long state,
 
 out:
+	for (i = 0; i < param_count; i++)
+		if (bufs[i])
```
```
	for (i = 0; i < param_count; i++)
		free_wipe(bufs[i]);
```

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -4259,8 +4400,9 @@ TEE_Result syscall_asymm_operate(unsigned long state,
-				       sizeof(hash));
+				res = copy_from_user(&hash,
+						     params[n].content.ref.buffer,
+						     sizeof(hash));
```
break on error

seonghp reviewed (commented):
Took another iteration. Still thinking about what would be proper implementation of `GET_USER` on platforms other than ARM64. Note that comments on `tee_svc_cryp.c` are mostly not visited yet.

Thank you for your precious review, and I am going to revisit the remaining comments, soon

etienne-lms commented on core/arch/arm/include/arm64.h:
```diff
@@ -460,6 +462,10 @@ DEFINE_REG_WRITE_FUNC_(icc_eoir0, uint32_t, S3_0_c12_c8_1)
 DEFINE_REG_WRITE_FUNC_(icc_igrpen0, uint32_t, S3_0_C12_C12_6)
 DEFINE_REG_WRITE_FUNC_(icc_igrpen1, uint32_t, S3_0_C12_C12_7)
+
+#define WRITE_PAN(val) \
```
please address @jenswi-linaro's commment

By the way, enabling/disabling PAN acts on current CPU or all CPUs?

etienne-lms commented on core/kernel/user_access.c:
```diff
@@ -29,8 +46,11 @@ TEE_Result copy_from_user(void *kaddr, const void *uaddr, size_t len)
 	res = check_access(flags, uaddr, len);
-	if (!res)
+	if (!res) {
+		enter_user_access();
```
could this be replaced by a call to `disable_pan()` which is more explicit than `enter_user_access()` regarding what the function is expected to do?

jforissier replied:
This is a non-arch-specific file so PAN should not be used IMHO.

etienne-lms replied:
It seems to me you said PAN was a generic feature, generic enough for `CFG_PAN` to cover all Privilege Access Never support :)

seonghp replied:
I'd prefer name that are architecture-agnostic, rather than that are specific to ARM architecture. RISC-V also has hardware feature called SSTATUS.SUM (which stands for 'permit Supervisor User Memory access') that is almost equivalent to PAN, and I believe architecture-agnostic naming shall be preferred so that RISC-V can reuse the interface by implementing that function in the future. Some names that I could came up with were `(dis)allow_user_access()` or `(dis)allow_unprivileged_access()`. Do you have more suggestions?

etienne-lms replied:
I would not use `user_access` because there is no user land access here. It's still core (privileged) access.
Proposition: `allow_core_user_map_accesss()` / `forbid_core_user_map_accesss()`?

jenswi-linaro replied:
I guess `enter_user_access()` should be read as "enter a section with user mode memory access temporarily enabled"

etienne-lms replied:
ok, you have the last word.

etienne-lms commented on core/kernel/user_access.c:
```diff
@@ -17,9 +17,26 @@
+
+		uctx = to_user_mode_ctx(s->ctx);
+		if (!uctx)
+			return TEE_ERROR_ACCESS_DENIED;
```
should be TEE_ERROR_GENERIC as this is not expected. even `panic()` seems appropriate to me.

etienne-lms commented on core/arch/arm/include/kernel/user_access_arch.h:
```diff
@@ -18,4 +18,26 @@ static inline void enter_user_access(void) {}
+#ifdef ARM64
+#define GET_USER(_x, _p) ({						\
+	TEE_Result __res = TEE_SUCCESS;					\
+	__typeof(_x) __x = 0;						\
```
any comment on this scalar/struct contraint on operand type?

seonghp replied:
I tried `typeof(*(_p)) __x = 0`, but it resulted in compiler warnings, due to the `const` on `_p`:
```
core/tee/tee_svc_cryp.c: In function ‘copy_in_attrs’:
core/arch/arm/include/kernel/user_access_arch.h:30:32: warning: passing argument 1 of ‘copy_from_user’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   30 |         __res = copy_from_user(&__x, (const void *)_p, sizeof(__x));    \
      |                                ^~~~
core/tee/tee_svc_cryp.c:1689:17: note: in expansion of macro ‘GET_USER’
 1689 |                 GET_USER(attrs[n].attributeID, &usr_attrs[n].attribute_id);
      |                 ^~~~~~~~
In file included from core/tee/tee_svc_cryp.c:14:
core/include/kernel/user_access.h:16:33: note: expected ‘void *’ but argument is of type ‘const uint32_t *’ {aka ‘const unsigned int *’}
   16 | TEE_Result copy_from_user(void *kaddr, const void *uaddr, size_t len);
      |                           ~~~~~~^~~~~
In file included from core/include/kernel/user_access.h:12,
                 from core/tee/tee_svc_cryp.c:14:
core/arch/arm/include/kernel/user_access_arch.h:30:32: warning: passing argument 1 of ‘copy_from_user’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   30 |         __res = copy_from_user(&__x, (const void *)_p, sizeof(__x));    \
      |                                ^~~~
core/tee/tee_svc_cryp.c:1694:25: note: in expansion of macro ‘GET_USER’
 1694 |                         GET_USER(a, &usr_attrs[n].a);
      |                         ^~~~~~~~
In file included from core/tee/tee_svc_cryp.c:14:
core/include/kernel/user_access.h:16:33: note: expected ‘void *’ but argument is of type ‘const uint64_t *’ {aka ‘const long long unsigned int *’}
   16 | TEE_Result copy_from_user(void *kaddr, const void *uaddr, size_t len);
      |                           ~~~~~~^~~~~
In file included from core/include/kernel/user_access.h:12,
                 from core/tee/tee_svc_cryp.c:14:
core/arch/arm/include/kernel/user_access_arch.h:30:32: warning: passing argument 1 of ‘copy_from_user’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   30 |         __res = copy_from_user(&__x, (const void *)_p, sizeof(__x));    \
      |                                ^~~~
core/tee/tee_svc_cryp.c:1695:25: note: in expansion of macro ‘GET_USER’
 1695 |                         GET_USER(b, &usr_attrs[n].b);
      |                         ^~~~~~~~
In file included from core/tee/tee_svc_cryp.c:14:
core/include/kernel/user_access.h:16:33: note: expected ‘void *’ but argument is of type ‘const uint64_t *’ {aka ‘const long long unsigned int *’}
   16 | TEE_Result copy_from_user(void *kaddr, const void *uaddr, size_t len);
      |                           ~~~~~~^~~~~
In file included from core/include/kernel/user_access.h:12,
                 from core/tee/tee_svc_cryp.c:14:
core/arch/arm/include/kernel/user_access_arch.h:30:32: warning: passing argument 1 of ‘copy_from_user’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   30 |         __res = copy_from_user(&__x, (const void *)_p, sizeof(__x));    \
      |                                ^~~~
core/tee/tee_svc_cryp.c:1705:25: note: in expansion of macro ‘GET_USER’
 1705 |                         GET_USER(buf, &usr_attrs[n].a);
      |                         ^~~~~~~~
In file included from core/tee/tee_svc_cryp.c:14:
core/include/kernel/user_access.h:16:33: note: expected ‘void *’ but argument is of type ‘const uint64_t *’ {aka ‘const long long unsigned int *’}
   16 | TEE_Result copy_from_user(void *kaddr, const void *uaddr, size_t len);
      |                           ~~~~~~^~~~~
```

seonghp replied:
Would it be acceptable to get rid of `const` from pointers that are used by `GET_USER` macro?

etienne-lms replied:
> I tried `typeof(*(_p)) __x = 0`, but it resulted in compiler warnings, due to the const on `_p`:

Indeed! I don't knwon how you can get rid of those `const` attributes.

Maybe you should discard my comment (apologies) and restore `typeof(_x) __x = 0` + add an assertion on `_x` and `*_p` types sizes.

By the way: `_x` and `__x` are too much alike, I would prefer you rename `__x` to something different, e.g. `__tmp`.

seonghp replied:
Thank you for your comments. I will revisit this, soon 😄

seonghp replied:
I rolled back the macro to use `typeof(*(_p))` and added compile-time size match check. Besides that, I had to change type of few variable in `tee_svc.c` and `tee_svc_cryp.c` to make compile-assertions to pass.

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -443,7 +468,104 @@ TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
+		goto out;
+
+	res = copy_to_user(dst, buf, num_bytes);
+out:
```
prefer instruction blocks instead of goto instruction when possible:
```c
	res = binh_copy_to(binh, (vaddr_t)buf, offs, num_bytes);
	if (!res)
		res = copy_to_user(dst, buf, num_bytes);

	free(buf);
	return res;

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -443,7 +468,104 @@ TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
+	if (res)
+		goto err_unmap_va;
+
+	res = vm_set_prot(uctx, _va, ROUNDUP(memsz, SMALL_PAGE_SIZE), TEE_MATTR_URW);
```
break line to fit into 80char/line max

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -516,20 +638,23 @@ TEE_Result ldelf_syscall_remap(unsigned long old_va, vaddr_t *new_va,
+	res = copy_to_user(dst, buf, num_bytes);
+out:
+	free(buf);
+	return res;
```
here also prefer not use a `goto` instruction.
```c
	res = crypto_rng_read(buf, num_bytes);
	if (!res)
		res = copy_to_user(dst, buf, num_bytes);

	free(buf);
	return res;

etienne-lms commented on core/kernel/user_ta.c:
```diff
@@ -54,7 +54,13 @@ static void init_utee_param(struct utee_params *up,
-	up->types = p->types;
+	/*
+	 * NOTE: The type of up->types is uint64_t, while the type of p->types
+	 * is uint32_t. Hence, to avoid unexpected behaviors
```
could you rephrase. It seems to me the sentence is unterminated.

etienne-lms commented on core/mm/vm.c:
```diff
@@ -554,6 +555,10 @@ static bool cmp_region_for_remap(const struct vm_region *r0,
 }
 
+/*
+ * NOTE: `new_va` always belongs to userland. Hence, it is safe to always use
```
If I'm right, 'hence' is not expected to be used to start a sentence. Prefer use 'therefore', unless someone argues i'm wrong :)

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3233,8 +3280,11 @@ static TEE_Result get_hkdf_params(uint32_t algo, const TEE_Attribute *params,
 				*salt_len = params[n].content.ref.length;
+				if (memdup_user(params[n].content.ref.buffer,
+						*salt_len, &bufs[n]))
+					return TEE_ERROR_OUT_OF_MEMORY;
```
should return `memdup_user()` return value?

jenswi-linaro commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -446,6 +447,91 @@ TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
+	if (vm_unmap(uctx, vaddr, memsz))
+		panic();
+
+	vm_set_ctx(uctx->ts_ctx);
```
No

jenswi-linaro commented on core/kernel/ts_manager.c:
```diff
@@ -62,7 +62,13 @@ struct ts_session *ts_get_calling_session(void)
 
 struct ts_session *ts_get_current_session_may_fail(void)
 {
-	return TAILQ_FIRST(&thread_get_tsd()->sess_stack);
```
Why doesn't this work any longer?

etienne-lms replied:
that's right. Why is added function `thread_get_tsd_may_fail()` needed? Commit message says:
> Previously, ts_get_current_session_may_fail was calling
thread_get_tsd(), which may abort if current thread_id is invalid.

thread_id is a core information, unrelated to userland mapping. Is is really an issue in the scope of this P-R?

jenswi-linaro commented on core/kernel/user_access.c:
```diff
@@ -17,9 +17,26 @@
+			return TEE_ERROR_ACCESS_DENIED;
+
+		uctx = to_user_mode_ctx(s->ctx);
+		if (!uctx)
```
`uctx` is **always** non-NULL

jenswi-linaro commented on core/mm/vm.c:
```diff
@@ -566,9 +571,14 @@ TEE_Result vm_remap(struct user_mode_ctx *uctx, vaddr_t *new_va, vaddr_t old_va,
 	struct vm_region *r_first = NULL;
 	struct fobj *fobj = NULL;
 	vaddr_t next_va = 0;
+	vaddr_t _new_va = 0;
```
Please use a proper name, the underscore prefix is reserved for other situations.

etienne-lms commented on core/arch/arm/include/kernel/user_access_arch.h:
```diff
@@ -23,4 +23,26 @@ static inline void enter_user_access(void) {}
+})
+#else
+#define GET_USER(_x, _p) ({ (_x) = *(_p); TEE_SUCCESS; })
+#define PUT_USER(_x, _p) ({ *(_p) = (_x); TEE_SUCCESS; })
```
These ARM32 macros definitly need to use `copy_from_user()`/`copy_to_user()` as later change remove access right check as expected done from `GET_USER()`/`PUT_USER()`.
Ditto for the RISCV versions of the macros.

etienne-lms commented on core/tee/tee_pobj.c:
```diff
@@ -64,14 +65,37 @@ TEE_Result tee_pobj_get(TEE_UUID *uuid, void *obj_id, uint32_t obj_id_len,
+		 * (See vm_check_access_rights), we know that if
+		 * copy_from_user(..., obj_id, ...) fails here,
+		 * it's because obj_id is not a user address. Hence, it is
+		 * now safe to assume that obj_id is in kernel address space.
```
It looks a bit fragile. Only `tee_pobj_get()` calls vm_check_access_rights() to check `obj_id`. And are we sure that future caller of  `tee_pobj_get()` will have checked that?

etienne-lms commented on core/include/kernel/user_access.h:
```diff
@@ -9,6 +9,7 @@
 #include <assert.h>
 #include <tee_api_types.h>
 #include <types_ext.h>
+#include <kernel/user_access_arch.h>
```
moive between lines 9 and 10.

seonghp replied:
updated

etienne-lms reviewed (commented):
Comments for commit "arm64: Introduce permissive PAN implementation".

etienne-lms reviewed (commented):
Comments for commit "core: add more user access functions":
Still some review comments not addressed.
In commit message, add parentheses to function labels:
> Add user access functions clear_user(), strnlen_user() and memdup_user(),

seonghp reviewed (commented):
Took another iteration. More fixes are to come. Thank you for your reviews @etienne-lms and @jenswi-linaro 😄

seonghp reviewed (commented):
Took another iteration. More fixes are to come. Thank you for your reviews @etienne-lms and @jenswi-linaro 😄

etienne-lms reviewed (commented):
For commit "uaccess: use get_user in save_panic_regs_a64_ta"

Add parentheses to function labels. Proposal:
> core: arm: uaccess: check user access in save_panic_regs_a64_ta()

Make a full sentence in commit message below the message header line, e.g.:
```patch
-as `pushed[]` belongs to the userland.
+Checks user access rights in save_panic_regs_a64_ta() as pushed[] array
+belongs to the userland.

etienne-lms reviewed (commented):
For commit "uaccess: use get_user in save_panic_regs_a64_ta"

Add parentheses to function labels. Proposal:
> core: arm: uaccess: check user access in save_panic_regs_a64_ta()

Make a full sentence in commit message below the message header line, e.g.:
```patch
-as `pushed[]` belongs to the userland.
+Checks user access rights in save_panic_regs_a64_ta() as pushed[] array
+belongs to the userland.

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -446,6 +447,91 @@ TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
+	binh = handle_lookup(&sys_ctx->db, handle);
+	if (!binh) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		return res;
```
`return TEE_ERROR_BAD_PARAMETERS;`

etienne-lms reviewed (commented):
Comments for commit "ldelf: Add map_zi_and_cp_from_bin system call".

etienne-lms commented on core/kernel/ldelf_loader.c:
```diff
@@ -356,26 +371,24 @@ TEE_Result ldelf_dlopen(struct user_mode_ctx *uctx, TEE_UUID *uuid,
+	dl_entry_arg.dlopen.flags = flags;
+
+	res = clear_user(arg, sizeof(*arg));
+	if (res)
```
is it really needed to clear `*arg` since the full `*arg` content is loaded by the next instruction?

seonghp replied:
oh yeah. I removed `clear_user()` call

etienne-lms commented on core/kernel/ldelf_loader.c:
```diff
@@ -408,28 +426,27 @@ TEE_Result ldelf_dlsym(struct user_mode_ctx *uctx, TEE_UUID *uuid,
+	dl_entry_arg->dlsym.symbol[len] = '\0';
+
+	res = copy_to_user(arg, dl_entry_arg, len_arg_with_sym);
+	if (res)
```
Can remove the `out` branch label by freeing `dl_entry_arg()` right after `copy_to_user()` returns.

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -51,11 +52,27 @@ TEE_Result ldelf_syscall_map_zi(vaddr_t *va, size_t num_bytes, size_t pad_begin,
+
+	res = GET_USER(vaddr, va);
+	if (res)
+		return res;
```
`mobj_put(mobj);` missing

seonghp replied:
moved up `GET_USER_SCALAR()` call

etienne-lms commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -133,17 +136,21 @@ TEE_Result ldelf_syscall_open_bin(const TEE_UUID *uuid, size_t uuid_size,
+	if (res)
+		return res;
+
 	binh = calloc(1, sizeof(*binh));
```
Is freeing `binh` missing in the error exit path?
If so, it is a bug unrelated to this commit and should be fixed seperatly (aka in a dedicated commit).

seonghp replied:
iiuc `binh` is freed within `binh_close()`?

etienne-lms commented on core/kernel/user_ta.c:
```diff
@@ -49,12 +50,26 @@
+	 * clear 8 bytes (`&up->types`) first, and fill-in 4 bytes copied from
+	 * `p->types` there.
+	 */
+	res = clear_user(&up->types, sizeof(up->types));
```
is this needed as the next instruction fully loads `up->types` content?

seonghp replied:
I was worried if the upper bytes of `up->types` may remain uninitialized, since the `sizeof(p->types)` is smaller than the `sizeof(up->types)`.

etienne-lms replied:
You'r right, size arg for `copy_to_user()` is `sizeof(p->types)`, not `sizeof(up->types)`.
Discard my comment.

seonghp replied:
ok, thanks!

etienne-lms reviewed (commented):
Comments for commit "uaccess: correctly apply user-access functions"

Please make a full sentence in the commit message below its header line.

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3839,15 +3962,22 @@ TEE_Result syscall_authenc_init(unsigned long state, const void *nonce,
 				  aad_len, payload_len);
 	if (res != TEE_SUCCESS)
-		return res;
+		goto out;
```
prefer:
```patch
	res = crypto_authenc_init(cs->ctx, cs->mode, (uint8_t *)(key + 1),
				  key->key_size, nonce_kbuf, nonce_len, tag_len,
				  aad_len, payload_len);
+	free(nonce_kbuf);
 	if (res != TEE_SUCCESS)
-		goto out;
+		return res;

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3785,27 +3905,29 @@ TEE_Result syscall_cryp_derive_key(unsigned long state,
 
 out:
+	for (i = 0; i < param_count; i++)
+		free_wipe(bufs[i]);
```
consider case `bufs` was not allocated:
```c
	if (bufs)
		for (i = 0; i < param_count; i++)
			free_wipe(bufs[i]);

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -1678,16 +1686,25 @@ static TEE_Result copy_in_attrs(struct user_ta_ctx *utc,
 
 	for (n = 0; n < attr_count; n++) {
-		attrs[n].attributeID = usr_attrs[n].attribute_id;
+		GET_USER(attrs[n].attributeID, &usr_attrs[n].attribute_id);
```
check return value

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -1678,16 +1686,25 @@ static TEE_Result copy_in_attrs(struct user_ta_ctx *utc,
+			uint64_t b = 0;
+
+			GET_USER(a, &usr_attrs[n].a);
+			GET_USER(b, &usr_attrs[n].b);
```
check return values

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -1678,16 +1686,25 @@ static TEE_Result copy_in_attrs(struct user_ta_ctx *utc,
 					 TEE_MEMORY_ACCESS_ANY_OWNER;
 
+			GET_USER(buf, &usr_attrs[n].a);
+			get_user_u64_as_size_t(&len, (void *)&usr_attrs[n].b);
```
check return values

etienne-lms reviewed (commented):
Comments for commit "uaccess: apply user-access fn's to crypto services":

There are many occurrences of call to `memdup_user()` that report `TEE_ERROR_OUT_OF_MEMORY` on the function failure whereas it could need to report `TEE_ERROR_ACCESS_DENIED`. All should be fixed. The TA can behave differently in thoses 2 error cases, starting from a TA panic that can be required by the GP TEE Internal core API specification:
```patch
-				if (memdup_user(params[n].content.ref.buffer,
-						*info_len, &bufs[n]))
-					return TEE_ERROR_OUT_OF_MEMORY;
+				res = memdup_user(params[n].content.ref.buffer,
+						  *info_len, &bufs[n]);
+				if (res)
+					return res;

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
 "core: Apply finer-grained PAN".

seonghp commented on core/kernel/ldelf_syscalls.c:
```diff
@@ -443,7 +468,104 @@ TEE_Result ldelf_syscall_copy_from_bin(void *dst, size_t offs, size_t num_bytes,
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	return binh_copy_to(binh, (vaddr_t)dst, offs, num_bytes);
+	buf = malloc(num_bytes);
```
This path is taken when there are segments that overlap. If that was the case, then it allocates a buffer of size `seg.filesz - (size of overlapped region)`, which may be large in certain circumstance.

That being said, I am not sure how common are such cases. While running `xtest`, there was only a TA which has overlapped segments. That case the size of the allocated buffer was roughly 6KiB.

jenswi-linaro replied:
Assuming that a 6KiB allocation from the heap always succeeds is optimistic. There are probably TAs out there that require much larger allocations. This must always work unless we've run out of heap completely, in such a case it's not due to the TA but rather tuning of the heap or even some memory leakage.

seonghp replied:
https://github.com/OP-TEE/optee_os/blob/88dfa32787dbec5a470b6595c7988c830cde99b0/ldelf/ta_elf.c#L793-L805

The allocation can be avoided using `sys_map_zi_and_copy_from_ta_bin()`. Yet the complication here is that the file handle's offset has already passed the `seg->offset`. I think I can come up with some workaround. I will submit change to this PR. Could you please take a look then?

jenswi-linaro replied:
Absolutely.

seonghp commented on core/arch/riscv/include/kernel/user_access_arch.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef KERNEL_USER_ACCESS_ARCH_H
+#define KERNEL_USER_ACCESS_ARCH_H
+
+static inline void enter_user_access(void) {}
```
updated

seonghp reviewed (commented):
Took another iteration. More fixes are to come. Thanks!

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3248,8 +3300,12 @@ static TEE_Result get_hkdf_params(uint32_t algo, const TEE_Attribute *params,
 				*info_len = params[n].content.ref.length;
+				if (memdup_user(params[n].content.ref.buffer,
+						*info_len, &bufs[n]))
+					return TEE_ERROR_OUT_OF_MEMORY;
```
return the error code returned by `memdup_user()`

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3291,8 +3348,12 @@ static TEE_Result get_concat_kdf_params(const TEE_Attribute *params,
 				*other_info_len = params[n].content.ref.length;
+				if (memdup_user(params[n].content.ref.buffer,
+						*other_info_len, &bufs[n]))
+					return TEE_ERROR_OUT_OF_MEMORY;
```
return the error code returned by `memdup_user()`

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3332,8 +3393,11 @@ static TEE_Result get_pbkdf2_params(const TEE_Attribute *params,
 				*salt_len = params[n].content.ref.length;
+				if (memdup_user(params[n].content.ref.buffer,
+						*salt_len, &bufs[n]))
+					return TEE_ERROR_OUT_OF_MEMORY;
```
return the error code returned by `memdup_user()`

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3409,47 +3474,75 @@ static TEE_Result get_sm2_kep_params(const TEE_Attribute *params,
-					     p->content.ref.length,
+			if (memdup_user(p->content.ref.buffer,
+					p->content.ref.length, &bufs[n]))
+				return TEE_ERROR_OUT_OF_MEMORY;
```
return the error code returned by `memdup_user()`

ditto at lines 3488, 3498, 3508 etc...

etienne-lms reviewed (commented):
looks almost good to me.
please address the remaining review comments.

seonghp commented:
Dumped wip. I will handle each comment, soon.

I removed `<kernel/user_access.h>` from `tee_pobj.c` and `tee_supp_plugin_rpc.c` to resolve build failure with `CFG_WITH_USER_TA=n`.

seonghp commented on ldelf/syscalls_asm.S:
```diff
@@ -5,16 +5,17 @@
+LDELF_SYSCALL	_ldelf_set_prot,		LDELF_SET_PROT,			4
+LDELF_SYSCALL	_ldelf_remap,			LDELF_REMAP,			7
+LDELF_SYSCALL	_ldelf_gen_rnd_num,		LDELF_GEN_RND_NUM,		2
+LDELF_SYSCALL	_ldelf_map_zi_and_cp_from_bin	LDELF_MAP_ZI_AND_CP_FROM_BIN	8
```
updated

seonghp commented on ldelf/ta_elf.c:
```diff
@@ -874,17 +874,15 @@ static void populate_segments(struct ta_elf *elf)
-							   elf->handle, offset);
-				if (res)
-					err(res, "sys_copy_from_ta_bin");
+					err(res, "sys_map_zi_and_copy_from_ta_bin");
```
updated

seonghp commented on core/kernel/ldelf_loader.c:
```diff
@@ -356,26 +367,37 @@ TEE_Result ldelf_dlopen(struct user_mode_ctx *uctx, TEE_UUID *uuid,
+		goto out;
 	}
 
-	memset(arg, 0, sizeof(*arg));
```
First, writing constant (i.e., `LDELF_DL_ENTRY_DLOPEN`) to user memory will likely to be ugly, requiring intermediate variable. Also, by calling `clear_user()` and `PUT_USER()` multiple times, it would end up calling `check_access()` multiple times, which i think is suboptimal

seonghp commented on core/kernel/ldelf_loader.c:
```diff
@@ -398,6 +411,11 @@ TEE_Result ldelf_dlopen(struct user_mode_ctx *uctx, TEE_UUID *uuid,
+ * to access them.
+ */
 TEE_Result ldelf_dlsym(struct user_mode_ctx *uctx, TEE_UUID *uuid,
 		       const char *sym, size_t maxlen, vaddr_t *val)
```
slightly reworked `ldelf_sym` function to take `sym_len` instead of `maxlen`.

seonghp commented on core/kernel/ldelf_loader.c:
```diff
@@ -408,28 +426,27 @@ TEE_Result ldelf_dlsym(struct user_mode_ctx *uctx, TEE_UUID *uuid,
-		EMSG("ldelf stack is inaccessible!");
-		return res;
-	}
+	dl_entry_arg = malloc(len_arg_with_sym);
```
avoided that problem by calling `copy_to_user` twice for `struct dl_entry_arg` and `sym`, separately.

etienne-lms reviewed (commented):
Using PAN in threaded context need the CPU PAN config to be part of the thread context.
Maybe helper functions like `thread_enter_user_access()` & `thread_exit_user_access()` would help.

seonghp commented on core/include/kernel/user_access.h:
```diff
@@ -34,6 +34,35 @@ static inline TEE_Result copy_from_user(void *kaddr __unused,
+ *          Note that the returned buffer is allocated by malloc() and
+ *          the caller shall free() the buffer later.
+ * @returns TEE_SUCCESS on success, and TEE_ERROR_OUT_OF_MEMORY or
+ *          TEE_ERROR_ACCESS_DENIEND on error.
```
updated

seonghp commented on core/include/kernel/user_access.h:
```diff
@@ -34,6 +34,35 @@ static inline TEE_Result copy_from_user(void *kaddr __unused,
+
+/*
+ * memdup_user_private() - duplicate a private user-space buffer.
+ * @src:    Pointer to the user buffer to be duplicated. The buffer shuold
```
updated

seonghp commented on core/include/kernel/user_access.h:
```diff
@@ -34,6 +34,35 @@ static inline TEE_Result copy_from_user(void *kaddr __unused,
+TEE_Result memdup_user(const void *src, size_t len, void **p);
+
+/*
+ * memdup_user_private() - duplicate a private user-space buffer.
```
updated

seonghp commented on core/include/kernel/user_access.h:
```diff
@@ -34,6 +34,35 @@ static inline TEE_Result copy_from_user(void *kaddr __unused,
+ *          Note that the returned buffer is allocated by malloc() and
+ *          the caller shall free() the buffer later.
+ * @returns TEE_SUCCESS on success, and TEE_ERROR_OUT_OF_MEMORY or
+ *          TEE_ERROR_ACCESS_DENIEND on error.
```
updated

seonghp commented on core/include/kernel/user_access.h:
```diff
@@ -34,6 +34,35 @@ static inline TEE_Result copy_from_user(void *kaddr __unused,
+size_t strnlen_user(const void *s, size_t n);
+
+/*
+ * memdup_user() - duplicate a user-space buffer.
```
updated

seonghp commented:
Thanks @etienne-lms! To my understanding, PAN config is part of PSTATE and it's kept within the thread context (`cpsr` field of `struct thread_ctx_regs`). Regarding the function name, would it be more natural to rename `enter_user_access()` to `thread_enter_user_access()` and move it to `thread_arch.h` or `thread_arch_private.h`?

jenswi-linaro commented on core/kernel/ldelf_loader.c:
```diff
@@ -398,38 +407,39 @@ TEE_Result ldelf_dlopen(struct user_mode_ctx *uctx, TEE_UUID *uuid,
-	arg->dlsym.uuid = *uuid;
-	memcpy(arg->dlsym.symbol, sym, len);
-	arg->dlsym.symbol[len] = '\0';
+	res = copy_to_user(&arg->dlsym.symbol, sym, sym_len + 1);
```
Please drop the `&` in front of `arg->dlsym.symbol`.

seonghp commented on ldelf/syscalls_asm.S:
```diff
@@ -15,6 +15,7 @@ LDELF_SYSCALL	_ldelf_open_bin,	LDELF_OPEN_BIN,		3
+LDELF_SYSCALL	_ldelf_set_prot,	LDELF_SET_PROT		4
 LDELF_SYSCALL	_ldelf_remap,		LDELF_REMAP,		7
 LDELF_SYSCALL	_ldelf_gen_rnd_num,	LDELF_GEN_RND_NUM,	2
+LDELF_SYSCALL	_ldelf_map_zi_and_cp_from_bin	LDELF_MAP_ZI_AND_CP_FROM_BIN 8
```
updated

seonghp commented on ldelf/syscalls_asm.S:
```diff
@@ -15,6 +15,7 @@ LDELF_SYSCALL	_ldelf_open_bin,	LDELF_OPEN_BIN,		3
 LDELF_SYSCALL	_ldelf_map_bin,		LDELF_MAP_BIN,		8
 LDELF_SYSCALL	_ldelf_cp_from_bin,	LDELF_CP_FROM_BIN,	4
-LDELF_SYSCALL	_ldelf_set_prot,	LDELF_SET_PROT,		4
+LDELF_SYSCALL	_ldelf_set_prot,	LDELF_SET_PROT		4
```
updated

seonghp commented on ldelf/syscalls_asm.S:
```diff
@@ -6,7 +6,7 @@
 
 /* macro	function name		syscall num		num of args */
-LDELF_SYSCALL	_ldelf_return,		LDELF_RETURN,		1
+LDELF_SYSCALL	_ldelf_return		LDELF_RETURN,		1
```
updated

etienne-lms commented:
>Thanks @etienne-lms! To my understanding, PAN config is part of PSTATE and it's kept within the thread context (cpsr field of struct thread_ctx_regs). 

You're right. It part of the thread context.

> Regarding the function name, would it be more natural to rename enter_user_access() to thread_enter_user_access() and move it to thread_arch.h or thread_arch_private.h?

No, not needed. Sorry for the noise. Fine as-in.

etienne-lms commented on core/arch/arm/include/kernel/user_access_arch.h:
```diff
@@ -0,0 +1,28 @@
+ * Copyright (c) 2023, Amazon.com Inc. or its affiliates. All rights Reserved.
+ */
+
+#ifndef KERNEL_USER_ACCESS_ARCH_H
```
nitpicking: prefer with `__` prefix to better match other recently updates .h header file in this dir.
```c
#ifndef __KERNEL_USER_ACCESS_ARCH_H
#define __KERNEL_USER_ACCESS_ARCH_H
(...)
#endif /*__KERNEL_USER_ACCESS_ARCH_H*/

seonghp replied:
updated

etienne-lms commented on core/kernel/user_access.c:
```diff
@@ -68,8 +90,79 @@ TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len)
+{
+	TEE_Result res = TEE_SUCCESS;
+
+	*p = malloc(len);
```
all good, but it would be bettre `*p` is unaffected upon failure.
suggestion below (also applies to  `memdup_user_private()`) but feel free to discard my comment:
```c
	TEE_Result res = TEE_SUCCESS;
	void *buf = NULL;

	buf = malloc(len);
	if (!buf)
		return TEE_ERROR_OUT_OF_MEMORY;

	res = copy_from_user(buf, src, len);
	if (res)
		free(buf);
	else
		*p = buf;

	return res;
```

seonghp replied:
updated both comment and code, thanks!

etienne-lms commented on core/kernel/user_ta.c:
```diff
@@ -159,9 +179,12 @@ static TEE_Result user_ta_enter(struct ts_session *session,
+		res = clear_user(usr_params, sizeof(*usr_params));
+
+	if (res)
+		goto out;
```
Suggestion:
```patch
-	if (res) {
-		ts_pop_current_session();
-		goto out;
+	if (res)
+		goto out_pop_session;
 
 	(...)
 
+out_pop_session:
 	ts_sess = ts_pop_current_session();
 	assert(ts_sess == session);
 out:
 	dec_recursion();
 	(...)
```

etienne-lms reviewed (commented):
few late comments

jenswi-linaro commented:
> Hi @jenswi-linaro, regarding your concerns about potential large memory allocation, I have changed the logic of `sys_map_zi_and_copy_from_bin()` a bit and made it be used in _`copy_remapped_to()`_ case as well. I think the code doesn't look very satisfactory at this moment. That said if you like this approach, I can improve readability, etc.

I'm not sure I get what you're doing yet. The troubling memory allocation is still there. It is a bit confusing with map_offset and file_offset, but I guess a few suitable comments should be able to sort out that.

seonghp commented:
Yes, the troubling memory allocation is still in `ldelf_syscall_copy_from_bin()` syscall, but it is now only called by:
1. `copy_section_headers()` (the size of section headers will not going to be that large), and
2. `populate_segments_legacy()` (which I think is not going to be triggered in most cases).

Previously, `ldelf_syscall_copy_from_bin()` was also called in `populate_segments()` when the segment is `remapped_writeable`, which happens when the segment has overlapped region with the previous segment, possibly allocating large buffers (e.g., 6KiB).

The change I have introduced is to use `sys_map_zi_and_copy_from_ta_bin()` instead of `sys_copy_from_ta_bin()` for the above case, in order to get rid of the bounce buffer allocation in the common path.

`map_offset` and `file_offset` specify the offset within the map where the file contents are going to be copied to, and the offset in the ELF file, where the contents will be read from, respectively. `map_offset` was added to handle the above case, where the first few bytes are already read from the handle, and thus should be copied from the memory. `map_offset` can be used to skip the head of the map (`map[0:map_offset]`) when copying data from the handle.

jenswi-linaro commented:
I'm not convinced of the PAN stuff in the ldlef syscalls. For the sake of this PR perhaps the best would be to excuse ldelf from PAN for now and do that in a follow-up PR instead. Thoughts?

jenswi-linaro commented:
> I think that will work. Can I remove changes to ldelf syscalls from this PR?

Yes, please.

jenswi-linaro commented:
> (edited)
> 
> Shall I squash those commits together?
> 
> * uaccess: use get_user in save_panic_regs_a64_ta
> * uaccess: correctly apply user-access functions
> * uaccess: apply user-access fn's to crypto services
> 
> Also, I'd like to move `enter_user_access()` and `exit_user_access()` pairs added to `tee_svc_cryp.c` to `core: Apply finer-grained PAN`. May I make this edit?

Sure. While you're at it, I believe @etienne-lms suggested renaming the enter exit functions to `thread_enter_user_access()` and `thread_exit_user_access() and also keep them together with the other thread_* functions.

seonghp commented:
@jenswi-linaro  I made an edit that I mentioned earlier, besides renaming `enter_user_access()` to `thread_enter_user_access()`. I thought @etienne-lms said that `enter_user_access()` looks fine as-is. Would you prefer renaming those functions?

jenswi-linaro commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3332,8 +3380,15 @@ static TEE_Result get_pbkdf2_params(const TEE_Attribute *params,
+				TEE_Result res = TEE_SUCCESS;
+
 				*salt_len = params[n].content.ref.length;
+				res = memdup_user(params[n].content.ref.buffer,
```
Define max size.

jenswi-linaro commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3409,47 +3465,83 @@ static TEE_Result get_sm2_kep_params(const TEE_Attribute *params,
 		case TEE_ATTR_ECC_PUBLIC_VALUE_X:
-			crypto_bignum_bin2bn(p->content.ref.buffer,
-					     p->content.ref.length,
+			res = memdup_user(p->content.ref.buffer,
```
Define max size.

jenswi-linaro commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3409,47 +3465,83 @@ static TEE_Result get_sm2_kep_params(const TEE_Attribute *params,
+			if (res)
+				return res;
+
+			crypto_bignum_bin2bn(bufs[n], p->content.ref.length,
```
We should free `bufs[n]` as soon as we're done with it to minimize heap utilization.

Same below for the other bignums.

jenswi-linaro reviewed (commented):
Comments for "uaccess: correctly apply user-access functions".

We have quite a few places with unbounded memory allocations, and overall I suspect that we increase the pressure on the heap quite a bit. I'm concerned that we may run into random heap allocation failures depending on TA workload. The Linux kernel deals with such problems by waiting for memory to become available, but I'm not sure that's an option for us, at least not as a short-term solution.

I believe we need another approach to bounce buffers. How about a separate preallocated bounce buffer area for each TA context? That way TAs can only mess up things for themselves and we don't need to worry too much about running out of memory when allocating bounce buffers. I can provide some code in another RFC PR.

seonghp commented:
@jenswi-linaro I agree that bounce buffer allocations may cause non-deterministic heap pressure, and reserving memory region for bounce buffers will be a good option to isolate or contain memory pressure. Yet I am not sure how to choose the size of such bounce buffer pool for now tbh.

IIUC you are going to create a new RFC PR with proposed scheme, am I right? I'd like it if you can provide help on this feature 😄

jenswi-linaro commented:
> IIUC you are going to create a new RFC PR with proposed scheme, am I right?

Yes, that's my plan. With enough code to show what I have in mind.

seonghp commented:
Closing this PR in favor of #6135
